pex/specifier_sets.py

```diff
@@ -0,0 +1,331 @@
+
+    def __contains__(self, other):
+        # type: (Any) -> bool
+        if Range is not type(other):
+            return NotImplemented
+        return self.contains(other)
+
+    def contains(self, other):
+        # type: (Range) -> bool
+
+        if self.lower:
+            if not other.lower:
+                return False
+            elif self.lower.version > other.lower.version:
+                return False
+            elif (
+                self.lower.version == other.lower.version
+                and not self.lower.inclusive
+                and other.lower.inclusive
+            ):
+                return False
+
+        if self.upper:
+            if not other.upper:
+                return False
+            elif self.upper.version < other.upper.version:
+                return False
+            elif (
+                self.upper.version == other.upper.version
+                and not self.upper.inclusive
+                and other.upper.inclusive
+            ):
+                return False
+
+        for excluded_range in self.excludes:
+            if excluded_range.intersects(other):
+                return False
+
+        return True
+
+    def __str__(self):
+        # type: () -> str
+        parts = []
+        if self.lower:
+            parts.append(str(self.lower))
+        for excluded_range in self.excludes:
+            parts.append("!({})".format(excluded_range))
+        if self.upper:
+            parts.append(str(self.upper))
+        return ",".join(parts)
+
+
+def _increment_release(
+    version,  # type: Version
+    index,  # type: int
+):
+    # type: (...) -> Version
+
+    parsed_version = version.parsed_version
+
+    parts = []  # type: List[str]
+    if parsed_version.epoch != 0:
+        parts.append(str(parsed_version.epoch))
+        parts.append("!")
+    if parsed_version.release:
+        ceiling_release = list(parsed_version.release[:index])
+        ceiling_release.append(parsed_version.release[index] + 1)
+        parts.append(".".join(str(x) for x in ceiling_release))
+    if parsed_version.pre is not None:
+        parts.append("".join(str(x) for x in parsed_version.pre))
+    if parsed_version.post is not None:
+        parts.append(".post")
+        parts.append(str(parsed_version.post))
+    if parsed_version.dev is not None:
+        parts.append(".dev")
+        parts.append(str(parsed_version.dev))
+    if parsed_version.local is not None:
+        parts.append("+")
+        parts.append(parsed_version.local)
+
+    return Version("".join(parts))
+
+
+def _bounds(specifier_set):
+    # type: (SpecifierSet) -> Iterator[Union[ArbitraryEquality, ExcludedRange, LowerBound, UpperBound]]
+
+    for spec in specifier_set:
+        if "===" == spec.operator:
+            yield ArbitraryEquality(spec.version)
+            continue
+
+        if spec.version.endswith(".*"):
+            if "~=" == spec.operator:
+                raise ValueError()
```

Comment:
Yup, thanks. Will fix.

Reply:
Actually the else comment below covers `~=` - its impossible to get a SpecifierSet object with this content, since it fails to parse / construct. I'll just kill this checking code.